### PR TITLE
Normative: Adding Intl.DisplayNames specification.

### PR DESCRIPTION
**Pull request by FrankYFTang**
Fixes https://github.com/tc39/ecma402/issues/31.

Based on https://github.com/tc39/proposal-intl-displaynames which is ready for Stage 4.

### CODE REVIEW

**Comment by FrankYFTang** [comment]
We need to address the issue @anba raised about  "Conformance" section in this PR
https://github.com/tc39/proposal-intl-displaynames/issues/78

File: spec/conventions.html [hunk 1]
```diff
@@ -66,6 +66,11 @@ <h1>Well-Known Intrinsic Objects</h1>
           <td>`Intl.DateTimeFormat.prototype`</td>
           <td>The initial value of the *"prototype"* data property of the intrinsic %DateTimeFormat% (<emu-xref href="#sec-intl.datetimeformat.prototype"></emu-xref>).</td>
         </tr>
+        <tr>
+          <td>%DisplayNamesPrototype%</td>
+          <td>`Intl.DisplayNames.prototype`</td>
+          <td>The `Intl.DisplayNames` constructor (<emu-xref href="#sec-intl-displaynames-constructor"></emu-xref>).</td>
```
Comment (leobalter) [inline spec/conventions.html:72]:
```suggestion
          <td>%DisplayNames%</td>
          <td>`Intl.DisplayNames`</td>
          <td>The `Intl.DisplayNames` constructor (<emu-xref href="#sec-intl-displaynames-constructor"></emu-xref>).</td>
```

File: spec/displaynames.html [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    <h1>Abstract Operations for DisplayNames Objects</h1>
+
+    <emu-clause id="sec-canonicalcodefordisplaynames" aoid="CanonicalCodeForDisplayNames">
+      <h1>CanonicalCodeForDisplayNames ( _type_, _code_ )</h1>
+      <p>
+        The CanonicalCodeForDisplayNames abstract operation is called with arguments _type_, and _code_. It verifies that the _code_ argument represents a well-formed code according to the _type_ argument and returns the case-regularized form of the _code_. The algorithm refers to <a href="https://www.unicode.org/reports/tr35/#Identifiers">UTS 35's Unicode Language and Locale Identifiers grammar</a>. The following steps are taken:
+      </p>
+      <emu-alg>
+        1. If _type_ is `"language"`, then
+          1. If _code_ does not matches the `unicode_language_id` production, throw a *RangeError* exceptipon.
```
Comment (leobalter) [inline spec/displaynames.html:15]:
```suggestion
        1. If _type_ is *"language"*, then
          1. If _code_ does not match the `unicode_language_id` production, throw a *RangeError* exception.
```

Following patterns with 262, wrap string representations with star and double quotes, e.g. `*"string"*` instead the backtick quotes.

File: spec/displaynames.html [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+          1. If IsStructurallyValidLanguageTag(_code_) is *false*, throw a *RangeError* exception.
+          1. Set _code_ to CanonicalizeUnicodeLocaleId(_code_).
+          1. Return _code_.
+        1. If _type_ is `"region"`, then
+          1. If _code_ does not matches the `unicode_region_subtag` production, throw a *RangeError* exceptipon.
```
Comment (leobalter) [inline spec/displaynames.html:20]:
```suggestion
        1. If _type_ is *"region"*, then
          1. If _code_ does not match the `unicode_region_subtag` production, throw a *RangeError* exception.
```

File: spec/displaynames.html [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+          1. Let _code_ be the result of mapping _code_ to upper case as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>.
+          1. Return _code_.
+        1. If _type_ is `"script"`, then
+          1. If _code_ does not matches the `unicode_script_subtag` production, throw a *RangeError* exceptipon.
```
Comment (leobalter) [inline spec/displaynames.html:24]:
```suggestion
        1. If _type_ is *"script"*, then
          1. If _code_ does not match the `unicode_script_subtag` production, throw a *RangeError* exception.
```

File: spec/displaynames.html [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+          1. Let _code_ be the result of mapping the first character in _code_ to upper case, and mapping the second, third and fourth character in _code_ to lower case, as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>.
+          1. Return _code_.
+        1. Assert: _type_ is `"currency"`.
+        1. If ! IsWellFormedCurrencyCode(_code_) is *false*, throw a *RangeError* exceptipon.
```
Comment (leobalter) [inline spec/displaynames.html:28]:
```suggestion
        1. Assert: _type_ is *"currency"*.
        1. If ! IsWellFormedCurrencyCode(_code_) is *false*, throw a *RangeError* exception.
```

File: spec/displaynames.html [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+        1. Return _code_.
+      </emu-alg>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-intl-displaynames-constructor">
+    <h1>The Intl.DisplayNames Constructor</h1>
+
+    <p>
+      The DisplayNames constructor is a standard built-in property of the Intl object.</p>
```
Comment (leobalter) [inline spec/displaynames.html:39]:
```suggestion
    <p>
      The DisplayNames constructor is a standard built-in property of the Intl object.
    </p>
```

File: spec/displaynames.html [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        1. Let _localeData_ be %DisplayNames%.[[LocaleData]].
+        1. Let _matcher_ be ? GetOption(_options_, `"localeMatcher"`, `"string"`, &laquo; `"lookup"`, `"best fit"` &raquo;, `"best fit"`).
+        1. Set _opt_.[[localeMatcher]] to _matcher_.
+        1. Let _r_ be ResolveLocale(%DisplayNames%.[[AvailableLocales]], _requestedLocales_, _opt_, %DisplayNames%.[[RelevantExtensionKeys]]).
+        1. Let _style_ be ? GetOption(_options_, `"style"`, `"string"`, &laquo; `"narrow"`, `"short"`, `"long"` &raquo;, `"long"`).
+        1. Set _displayNames_.[[Style]] to _style_.
+        1. Let _type_ be ? GetOption(_options_, `"type"`, `"string"`, &laquo; `"language"`, `"region"`, `"script"`, `"currency"` &raquo;, *undefined*).
+        1. If _type_ is *undefined*, throw a *TypeError* exception.
+        1. Set _displayNames_.[[Type]] to _type_.
+        1. Let _fallback_ be ? GetOption(_options_, `"fallback"`, `"string"`, &laquo; `"code"`, `"none"` &raquo;, `"code"`).
```
Comment (leobalter) [inline spec/displaynames.html:63]:
```suggestion
        1. Let _displayNames_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%DisplayNames.prototype%"*, &laquo; [[InitializedDisplayNames]], [[Locale]], [[Style]], [[Type]], [[Fallback]], [[Fields]] &raquo;).
        1. Let _requestedLocales_ be ? CanonicalizeLocaleList(_locales_).
        1. Let _options_ be ? ToObject(_options_).
        1. Let _opt_ be a new Record.
        1. Let _localeData_ be %DisplayNames%.[[LocaleData]].
        1. Let _matcher_ be ? GetOption(_options_, *"localeMatcher"*, *"string"*, &laquo; *"lookup"*, *"best fit"* &raquo;, *"best fit"*).
        1. Set _opt_.[[localeMatcher]] to _matcher_.
        1. Let _r_ be ResolveLocale(%DisplayNames%.[[AvailableLocales]], _requestedLocales_, _opt_, %DisplayNames%.[[RelevantExtensionKeys]]).
        1. Let _style_ be ? GetOption(_options_, *"style"*, *"string"*, &laquo; *"narrow"*, *"short"*, *"long"* &raquo;, *"long"*).
        1. Set _displayNames_.[[Style]] to _style_.
        1. Let _type_ be ? GetOption(_options_, *"type"*, *"string"*, &laquo; *"language"*, *"region"*, *"script"*, *"currency"* &raquo;, *undefined*).
        1. If _type_ is *undefined*, throw a *TypeError* exception.
        1. Set _displayNames_.[[Type]] to _type_.
        1. Let _fallback_ be ? GetOption(_options_, *"fallback"*, *"string"*, &laquo; *"code"*, *"none"* &raquo;, *"code"*).
```

File: spec/displaynames.html [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+    <p>
+      The Intl.DisplayNames constructor has the following properties:
+    </p>
+
+    <emu-clause id="sec-Intl.DisplayNames.prototype">
+      <h1>Intl.DisplayNames.prototype</h1>
+
+      <p>
+        The value of *Intl.DisplayNames.prototype* is *%DisplayNamesPrototype%*.
```
Comment (leobalter) [inline spec/displaynames.html:91]:
```suggestion
        The value of `Intl.DisplayNames.prototype` is *%DisplayNames.prototype%*.
```

File: spec/displaynames.html [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        It is recommended that implementations use the locale data provided by the Common Locale Data Repository (available at <a href="http://cldr.unicode.org/">http://cldr.unicode.org/</a>).
+      </emu-note>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-properties-of-intl-displaynames-prototype-object">
+    <h1>Properties of the Intl.DisplayNames Prototype Object</h1>
+
+    <p>
+      The Intl.DisplayNames prototype object, referred to as *%DisplayNamesPrototype%*, is itself an ordinary object. It is not a Intl.DisplayNames instance, does not have an [[InitializedDisplayNames]] internal slot or any of the other internal slots of Intl.DisplayNames instance objects.
```
Comment (leobalter) [inline spec/displaynames.html:146]:
```suggestion
      The Intl.DisplayNames prototype object is itself an ordinary object. <dfn>%DisplayNames.prototype%</dfn> is not an Intl.DisplayNames instance and does not have an [[InitializedDisplayNames]] internal slot or any of the other internal slots of Intl.DisplayNames instance objects.
```

File: spec/displaynames.html [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+      <p>
+        The initial value of *Intl.DisplayNames.prototype.constructor* is the intrinsic object *%DisplayNames%*.
+      </p>
+    </emu-clause>
+
+    <emu-clause id="sec-Intl.DisplayNames.prototype-@@tostringtag">
+      <h1>Intl.DisplayNames.prototype[ @@toStringTag ]</h1>
+
+      <p>
+        The initial value of the @@toStringTag property is the string value `"Intl.DisplayNames"`.
```
Comment (leobalter) [inline spec/displaynames.html:161]:
```suggestion
        The initial value of the @@toStringTag property is the string value *"Intl.DisplayNames"*.
```

File: spec/displaynames.html [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        <li>The value of fields `"language"`, `"region"`, `"script"`, and `"currency"` must be a Records which must have fields with the names of one of the valid display name styles: `"narrow"`, `"short"`, and `"long"`.</li>
+        <li>The display name styles fields under display name type `"language"` should contain Records, with keys corresponding to language codes according to `unicode_language_id` production. The value of these fields must be string values.</li>
+        <li>The display name styles fields under display name type `"region"` should contain Records, with keys corresponding to region codes. The value of these fields must be string values.</li>
+        <li>The display name styles fields under display name type `"script"` should contain Records, with keys corresponding to script codes. The value of these fields must be string values.</li>
+        <li>The display name styles fields under display name type `"currency"` should contain Records, with keys corresponding to currency codes. The value of these fields must be string values.</li>
```
Comment (leobalter) [inline spec/displaynames.html:133]:
```suggestion
        <li>[[LocaleData]].[[&lt;_locale_&gt;]] must have a [[types]] field for all locale values _locale_. The value of this field must be a Record, which must have fields with the names of one of the valid display name types: *"language"*, *"region"*, *"script"*, and *"currency"*.</li>
        <li>The value of fields *"language"*, *"region"*, *"script"*, and *"currency"* must be a Records which must have fields with the names of one of the valid display name styles: *"narrow"*, *"short"*, and *"long"*.</li>
        <li>The display name styles fields under display name type *"language"* should contain Records, with keys corresponding to language codes according to `unicode_language_id` production. The value of these fields must be string values.</li>
        <li>The display name styles fields under display name type *"region"* should contain Records, with keys corresponding to region codes. The value of these fields must be string values.</li>
        <li>The display name styles fields under display name type *"script"* should contain Records, with keys corresponding to script codes. The value of these fields must be string values.</li>
        <li>The display name styles fields under display name type *"currency"* should contain Records, with keys corresponding to currency codes. The value of these fields must be string values.</li>
```

File: spec/displaynames.html [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+      <emu-alg>
+        1. Let _displayNames_ be *this* value.
+        1. If Type(_displayNames_) is not Object, throw a *TypeError* exception.
+        1. If _displayNames_ does not have an [[InitializedDisplayNames]] internal slot, throw a *TypeError* exception.
+        1. Let _code_ be ? ToString(_code_).
+        1. Let _code_ be ? CanonicalCodeForDisplayNames(_displayNames_.[[Type]], _code_).
+        1. Let _fields_ be _displayNames_.[[Fields]].
+        1. Let _name_ be _fields_[[&lt;_code_&gt;]].
+        1. If _name_ is not *undefined*, return _name_.
+        1. If _displayNames_.[[Fallback]] is `"code"`, return _code_.
```
Comment (leobalter) [inline spec/displaynames.html:184]:
```suggestion
        1. If _displayNames_.[[Fallback]] is *"code"*, return _code_.
```

File: spec/displaynames.html [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+            <td>[[Style]]</td>
+            <td>`"style"`</td>
+          </tr>
+          <tr>
+            <td>[[Type]]</td>
+            <td>`"type"`</td>
+          </tr>
+          <tr>
+            <td>[[Fallback]]</td>
+            <td>`"fallback"`</td>
```
Comment (leobalter) [inline spec/displaynames.html:232]:
```suggestion
            <td>*"locale"*</td>
          </tr>
          <tr>
            <td>[[Style]]</td>
            <td>*"style"*</td>
          </tr>
          <tr>
            <td>[[Type]]</td>
            <td>*"type"*</td>
          </tr>
          <tr>
            <td>[[Fallback]]</td>
            <td>*"fallback"*</td>
```

File: spec/displaynames.html [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        </table>
+      </emu-table>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-properties-of-intl-displaynames-instances">
+    <h1>Properties of Intl.DisplayNames Instances</h1>
+
+    <p>
+      Intl.DisplayNames instances are ordinary objects that inherit properties from %DisplayNamesPrototype%.
```
Comment (leobalter) [inline spec/displaynames.html:243]:
```suggestion
      Intl.DisplayNames instances are ordinary objects that inherit properties from %DisplayNames.prototype%.
```

File: spec/displaynames.html [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+
+    <p>
+      Intl.DisplayNames instances also have several internal slots that are computed by the constructor:
+    </p>
+
+    <ul>
+      <li>[[Locale]] is a String value with the language tag of the locale whose localization is used for formatting.</li>
+      <li>[[Style]] is one of the String values `"narrow"`, `"short"`, or `"long"`, identifying the display names style used.</li>
+      <li>[[Type]] is one of the String values `"language"`, `"region"`, `"script"`, or `"currency"`, identifying the type of the display names requested.</li>
+      <li>[[Fallback]] is one of the String values `"code"`, or `"none"`, identifying the fallback return when the system does not have the requested display name.</li>
```
Comment (leobalter) [inline spec/displaynames.html:258]:
```suggestion
      <li>[[Style]] is one of the String values *"narrow"*, *"short"*, or *"long"*, identifying the display name style used.</li>
      <li>[[Type]] is one of the String values *"language"*, *"region"*, *"script"*, or *"currency"*, identifying the type of the display names requested.</li>
      <li>[[Fallback]] is one of the String values *"code"*, or *"none"*, identifying the fallback return when the system does not have the requested display name.</li>
```

File: spec/displaynames.html [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+      <li>[[Fields]] is a Record (see <emu-xref href="#sec-Intl.DisplayNames-internal-slots"></emu-xref>) which
+        must have fields with keys corresponding to codes according to [[Style]] and [[Type]].</li>
```
Comment (leobalter) [inline spec/displaynames.html:260]:
```suggestion
      <li>[[Fields]] is a Record (see <emu-xref href="#sec-Intl.DisplayNames-internal-slots"></emu-xref>) which must have fields with keys corresponding to codes according to [[Style]] and [[Type]].</li>
```

**Review by leobalter — CHANGES_REQUESTED** [review]

**Comment by leobalter** [comment]
As my review got too verbose, I'd like to bring back a pending feedback from @anba:

> https://tc39.es/ecma402/#conformance lists options which are allowed to have implementation-defined behaviour. For Intl.DisplayNames, the "style" and "type" options should probably be listed there. I'm not yet sure if it makes sense to allow implementation-defined behaviour for the "fallback" option, because I don't exactly know what other kind of fallback should happen here.

@FrankYFTang I've got some minor suggestions, all editorial which I can apply (probably tomorrow). 

I can also help on the missing conformance part, please let me know!

**Comment by FrankYFTang** [comment]
> @FrankYFTang I've got some minor suggestions, all editorial which I can apply (probably tomorrow).


Thanks. Great ! 

> 
> I can also help on the missing conformance part, please let me know!

That will be nice.

**Review by leobalter — APPROVED** [review]

**Comment by leobalter** [comment]
Ok, the conflicts are resolved and I'll take a look at this conformance part. I believe it's best to do in a follow up PR we work through this week.

**Comment by FrankYFTang** [comment]
> Ok, the conflicts are resolved and I'll take a look at this conformance part. I believe it's best to do in a follow up PR we work through this week.

Agree

**Pull request merged**